File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/c/acquire-lock.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.AcquireLockInput;
type Output = DistributedCache.AcquireLockOutput;

export class AcquireLock
  extends BaseHandler<Input, Output>
  implements DistributedCache.IAcquireLockHandler
{
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    if (!Number.isFinite(input.expirationMs) || input.expirationMs <= 0) {
      return D2Result.validationFailed({
        inputErrors: [["expirationMs", "Must be a finite positive number."]],
      });
    }

    try {
      const wasSet = await this.redis.set(input.key, input.lockId, "PX", input.expirationMs, "NX");
      return D2Result.ok({ data: { acquired: wasSet === "OK" } });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}

File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/C/AcquireLock.cs
// -----------------------------------------------------------------------
// <copyright file="AcquireLock.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.C;

using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using H = D2.Shared.Interfaces.Caching.Distributed.Handlers.C.ICreate.IAcquireLockHandler;
using I = D2.Shared.Interfaces.Caching.Distributed.Handlers.C.ICreate.AcquireLockInput;
using O = D2.Shared.Interfaces.Caching.Distributed.Handlers.C.ICreate.AcquireLockOutput;

/// <summary>
/// Handler for acquiring a distributed lock in Redis using SET NX with mandatory TTL.
/// </summary>
public partial class AcquireLock : BaseHandler<H, I, O>, H
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="AcquireLock"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public AcquireLock(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            var db = r_redis.GetDatabase();
            var acquired = await db.StringSetAsync(
                input.Key,
                input.LockId,
                input.Expiration,
                When.NotExists);

            return D2Result<O?>.Ok(new O(acquired));
        }
        catch (RedisException ex)
        {
            LogAcquireLockFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<O?>.ServiceUnavailable(traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while acquiring a distributed lock.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while acquiring lock for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogAcquireLockFailed(ILogger logger, Exception ex, string key, string? traceId);
}
